### PR TITLE
fix(auth): strip surrounding whitespace from password fields

### LINKED ISSUE #481: fix(auth): password fields don't trim surrounding whitespace, causing false login failures
## Describe the Bug

The admin login form (and the regular user login/registration/password forms) do not trim leading or trailing whitespace from the submitted password before comparing or hashing it.
If a copied password picks up a stray space or trailing newline (e.g. copying ADMIN_PASSWORD=... out of a .env file, or copying from a terminal/password manager), the login silently fails with "Invalid password" even though the visible password is correct. 
Since the field is masked, there is no way for the user to notice the extra whitespace.

## Steps to Reproduce

1. Go to /admin/login.
2. Copy the admin password including a trailing space or newline (e.g. select the whole line ADMIN_PASSWORD=xxxx from .env and paste it, or just add a trailing space by hand).
3. Paste it into the password field and click "Sign In".
4. See error: "Invalid password", even though the password itself is correct.

## Expected Behavior
Leading/trailing whitespace around the password should be ignored, and the login should succeed, matching the behavior of most login forms (GitHub, Google, AWS console, etc.).

## Screenshots / Console Logs

https://github.com/user-attachments/assets/ac360770-7b30-4d98-99b5-b3ab462fadec

## Environment:
- OS: Linux
- Browser: Chrome
- Version: latest

## Additional Context
Root cause: admin_login_submit in portal/routers/admin.py does an exact password != settings.admin_password string comparison with no .strip(). The same missing-trim pattern also affects the regular user auth flow in portal/routers/auth.py (login, registration, set-password, reset-password, remove-password), since none of those strip whitespace either before hashing/verifying via hash_password / verify_password in portal/auth.py.

### PR DESCRIPTION
**Pull request by arpan7sarkar**
## Description

Password fields don't trim leading/trailing whitespace before comparing or hashing, so a stray space or newline from a copy-paste (e.g. copying `ADMIN_PASSWORD=...` out of `.env`) makes an otherwise-correct password fail silently.

## Related Issue

Closes #481

## Changes Made

- `admin_login_submit` strips both the submitted password and `settings.admin_password` before comparing, and checks emptiness on the trimmed value so a whitespace-only `ADMIN_PASSWORD` can't be satisfied by an empty submission.
- `hash_password` strips before hashing, so new hashes are normalized.
- `verify_password` tries the normalized candidate first, then falls back to the raw candidate for hashes created before this change (from a password with meaningful surrounding whitespace) — otherwise those users would be locked out with no way back in except a password reset. The login and remove-password handlers pass the raw value through so that fallback can run.
- `register_submit` rejects a whitespace-only password with an error instead of silently creating a passwordless account.

## Testing

Reproduced end-to-end with a real browser (Playwright) against `/admin/login`: a password with a trailing space returned "Invalid password" before the fix, and logs in after.

- [x] Tested locally (Playwright E2E repro, before/after)
- [x] Existing tests pass (`uv run pytest tests/ -v` — 564 passed)
- [x] Added tests: whitespace-tolerant admin/user login, `verify_password`, the whitespace-only-`ADMIN_PASSWORD` guard, and legacy-hash backward compatibility

https://github.com/user-attachments/assets/8e1e264e-6807-48a6-9d50-2b213457a90d

## Checklist

- [x] My code follows the project's coding standards
- [x] I have tested my changes
- [x] I have updated documentation where necessary (none needed — internal behavior fix, no API/doc surface changed)
- [x] My changes do not introduce new warnings or errors (`ruff check` clean on changed files)

## Summary by Sourcery

Normalize password whitespace handling across authentication flows while preserving compatibility with legacy password hashes.

Bug Fixes:
- Normalize surrounding whitespace in admin and user password handling so copied passwords authenticate consistently.
- Reject whitespace-only passwords during registration and prevent whitespace-only admin configuration from allowing empty-password login.
- Preserve compatibility with existing password hashes that include meaningful surrounding whitespace.

Tests:
- Add coverage for whitespace-tolerant admin login, whitespace-only admin password configuration, and password normalization behavior.

## Summary by CodeRabbit

- **Bug Fixes**
  - Admin login now handles accidental leading or trailing whitespace in submitted passwords.
  - Passwords for accounts preserve intentional whitespace instead of silently removing it.
  - Blank or whitespace-only passwords are rejected during registration and password updates.
  - Existing users can continue signing in with passwords created under the previous whitespace-handling behavior.

- **Tests**
  - Added coverage for admin login whitespace handling and invalid whitespace-only administrator passwords.

### CODE REVIEW

File: portal/auth.py [hunk 1]
```diff
@@ -20,11 +20,11 @@
 
 
 def hash_password(password: str) -> str:
-    return bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
+    return bcrypt.hashpw(password.strip().encode(), bcrypt.gensalt()).decode()
 
 
 def verify_password(password: str, password_hash: str) -> bool:
-    return bcrypt.checkpw(password.encode(), password_hash.encode())
+    return bcrypt.checkpw(password.strip().encode(), password_hash.encode())
```
[the remark this answers is not in the document]
Comment (arpan7sarkar) [reply portal/auth.py]: Fixed in a26e98a: verify_password now tries the normalized candidate first and falls back to the raw candidate when stripping changed it, so legacy hashes built from an unstripped password still verify. The login and remove-password handlers were updated to pass the raw submitted value through (they were stripping before the call, which would have made the fallback dead code).

[the remark this answers is not in the document]
Comment (arpan7sarkar) [reply portal/auth.py]: Fixed in a26e98a — same fix as suggested. verify_password normalizes first and falls back to the raw value for legacy hashes; user_login_submit and remove_password_api no longer strip before calling it, and there's a regression test (test_verify_accepts_legacy_hash_built_from_untrimmed_password).

File: portal/auth.py [hunk 2]
```diff
@@ -20,11 +20,18 @@
 
 
 def hash_password(password: str) -> str:
-    return bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
+    return bcrypt.hashpw(password.strip().encode(), bcrypt.gensalt()).decode()
```
Comment (ArnavBallinCode) [inline portal/auth.py]: Please revert this change. Modifying user passwords before hashing violates OWASP security guidelines against silent truncation. Passwords must be hashed exactly as submitted.

File: portal/auth.py [hunk 3]
```diff
@@ -20,11 +20,18 @@
...
 
 
 def verify_password(password: str, password_hash: str) -> bool:
-    return bcrypt.checkpw(password.encode(), password_hash.encode())
+    stripped = password.strip()
```
Comment (ArnavBallinCode) [inline portal/auth.py]: Please revert the changes to this function. We should not attempt to strip whitespace during password verification for standard user accounts.

File: portal/routers/auth.py [hunk 4]
```diff
@@ -113,13 +113,16 @@ async def register_submit(request: Request):
     form = await request.form()
     email = form.get("email", "").strip().lower()
     display_name = form.get("display_name", "").strip()
-    password = form.get("password", "")
+    password_raw = form.get("password", "")
+    password = password_raw.strip()
```
Comment (ArnavBallinCode) [inline portal/routers/auth.py:117]: Please revert this. We should hash the raw password. The validation below (`if password_raw and not password`) is good and can be kept to reject whitespace-only passwords, but we should pass `password_raw` to `create_user`.

File: portal/routers/auth.py [hunk 5]
```diff
@@ -343,7 +348,7 @@ async def reset_password_page(request: Request, token: str):
 @router.post("/auth/reset/{token}")
 async def reset_password_submit(request: Request, token: str):
     form = await request.form()
-    password = form.get("password", "")
+    password = form.get("password", "").strip()
```
Comment (ArnavBallinCode) [inline portal/routers/auth.py]: Please revert the `.strip()` here to ensure we don't truncate user passwords.

File: portal/routers/auth.py [hunk 6]
```diff
@@ -444,7 +449,7 @@ async def account_page(request: Request):
 async def set_password_api(request: Request):
     user_token = await require_user(request)
     data = await request.json()
-    password = data.get("password", "")
+    password = data.get("password", "").strip()
```
Comment (ArnavBallinCode) [inline portal/routers/auth.py]: Please revert the `.strip()` here as well.

File: tests/test_user_auth.py [hunk 7]
```diff
@@ -66,6 +67,19 @@ async def test_wrong_password_fails(self, setup_db):
         hashed = hash_password("correct")
         assert not verify_password("wrong", hashed)
 
+    @pytest.mark.anyio
+    async def test_verify_ignores_surrounding_whitespace(self, setup_db):
```
Comment (ArnavBallinCode) [inline tests/test_user_auth.py]: Please remove this test once `verify_password` is reverted.

File: tests/test_user_auth.py [hunk 8]
```diff
@@ -66,6 +67,19 @@ async def test_wrong_password_fails(self, setup_db):
...
+        hashed = hash_password("securepass123")
+        assert verify_password(" securepass123\n", hashed)
+
+    @pytest.mark.anyio
```
Comment (ArnavBallinCode) [inline tests/test_user_auth.py]: Please remove this legacy fallback test as well.

File: tests/test_user_auth.py [hunk 9]
```diff
@@ -161,6 +175,21 @@ async def test_login_with_correct_credentials(self, setup_db):
         assert resp.headers["location"] == "/account"
         assert "user_token" in resp.headers.get("set-cookie", "")
 
+    @pytest.mark.anyio
+    async def test_login_strips_surrounding_whitespace(self, setup_db):
```
Comment (ArnavBallinCode) [inline tests/test_user_auth.py]: Please remove this test once login trimming is reverted.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
Thank you for the PR! Fixing the `.env` copy-paste issue for the admin password is a great catch. However, modifying `hash_password` and `verify_password` to silently strip whitespace for *all* user accounts violates OWASP security guidelines. Passwords entered via web forms must be treated as exact binary data. If a user intentionally uses a password with a space, we should not truncate it behind their back.

Please revert the changes to `portal/auth.py`, `portal/routers/auth.py` (except the new validation check), and the related tests in `tests/test_user_auth.py`. Keep the excellent fix in `admin.py`!

**Comment by arpan7sarkar** [comment]
> Thank you for the PR! Fixing the `.env` copy-paste issue for the admin password is a great catch. However, modifying `hash_password` and `verify_password` to silently strip whitespace for _all_ user accounts violates OWASP security guidelines. Passwords entered via web forms must be treated as exact binary data. If a user intentionally uses a password with a space, we should not truncate it behind their back.
> 
> Please revert the changes to `portal/auth.py`, `portal/routers/auth.py` (except the new validation check), and the related tests in `tests/test_user_auth.py`. Keep the excellent fix in `admin.py`!

Got it @ArnavBallinCode  thanks for the review will remove those and update the pr asap

**Review by arpan7sarkar — COMMENTED** [review]
Hi @ArnavBallinCode  have a look i have updated the whole thing

**Review by Saksham-Sirohi — COMMENTED** [review]
HEAD addresses Arnav's 2026-09-16 CHANGES: portal/auth.py hash/verify no longer strip; register hashes password_raw and only rejects whitespace-only via strip check; admin login strips both submitted and configured ADMIN_PASSWORD and rejects whitespace-only admin secret (regression tests present). No human re-approval after the fix commit 6ff6f5dcbd (2026-09-17). Looks ready from my side; still needs another human approval before merge.

**Review by Nirupon123 — APPROVED** [review]
Thank you for the contribution! I’ve reviewed the PR thoroughly, including the authentication logic, edge cases, tests, and security considerations. Everything looks solid and is in line with our repository standards. The changes are well implemented and properly tested, so I’m happy to approve this PR and recommend merging it @ArnavBallinCode @Saksham-Sirohi .
